lib/puppet/resource_api.rb

```diff
@@ -99,12 +99,33 @@ def my_provider
               defaultto options[:default]
             end
 
-            case options[:type]
-            when 'String'
+            type = Puppet::Pops::Types::TypeParser.singleton.parse(options[:type])
+            validate do |value|
+              return true if type.instance?(value)
+
+              inferred_type = Puppet::Pops::Types::TypeCalculator.infer_set(value)
+              if inferred_type.is_a? Puppet::Pops::Types::PStringType
```

Comment:
`TypeCalculator#infer_set` can be rather expensive and the `inferred_type` is not really of any use unless there's a mismatch so I'd suggest using:

    if value.is_a?(String)

instead and move the call to `#infer_set` down to where the mismatch error is generated.

Reply:
Thanks, that makes sense.